dbos/_sys_db.py

```diff
@@ -1882,6 +1883,86 @@ def check_connection(self) -> None:
             dbos_logger.error(f"Error connecting to the DBOS system database: {e}")
             raise
 
+    @db_retry()
+    def write_stream(self, workflow_uuid: str, key: str, value: Any) -> None:
+        """
+        Write a key-value pair to the stream at the first unused offset.
+        Uses SERIALIZABLE isolation level and automatically retries on serialization errors.
+        """
+        if self._debug_mode:
+            raise Exception("called write_stream in debug mode")
+
+        while True:
+            try:
+                # Use SERIALIZABLE isolation level for consistency
```

Comment:
Why do we need `SERIALIZABLE`? A workflow will only have one running instance, it's deterministic, and this stream is append-only.

Reply:
To absolutely guarantee that records are inserted in incrementing sequential order, even in degenerate cases where a workflow is executing multiple times concurrently.

Reply:
Changed the implementation a bit so it isn't needed